// gitex-flow/build/test/ConventionalCommits.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const Utils_1 = require("../src/tools/Utils");
describe('Test ConventionCommits implementation', function () {
    it('should parse complex conventional commit messages', function () {
        return __awaiter(this, void 0, void 0, function* () {
            const logs = yield Utils_1.Utils.parseConventionalCommits([
                `fix(service): Removed support of unencrypted HTTP protocol
      
      This unencrypted protocol has led to several vulnerabilities in the framework.
      
      SECURITY: Only encrypted protocols are now allowed.
      REMOVED: Removed support for HTTP.
      BREAKING CHANGE: Removed HTTP endpoint in web service.
      
      closes #941, refs #1094, #1100`,
            ]);
            chai_1.assert.equal(logs[0].type, 'fix');
            chai_1.assert.equal(logs[0].scope, 'service');
            chai_1.assert.equal(logs[0].subject, 'Removed support of unencrypted HTTP protocol');
            chai_1.assert.isNull(logs[0].merge);
            chai_1.assert.deepStrictEqual(logs[0].notes, [
                {
                    title: 'SECURITY',
                    text: 'Only encrypted protocols are now allowed.',
                },
                {
                    title: 'REMOVED',
                    text: 'Removed support for HTTP.',
                },
                {
                    title: 'BREAKING CHANGE',
                    text: 'Removed HTTP endpoint in web service.\n      ',
                },
            ]);
            chai_1.assert.deepStrictEqual(logs[0].references, [
                {
                    action: 'closes',
                    owner: null,
                    repository: null,
                    issue: '941',
                    raw: '#941',
                    prefix: '#',
                },
                {
                    action: 'refs',
                    owner: null,
                    repository: null,
                    issue: '1094',
                    raw: '#1094',
                    prefix: '#',
                },
                {
                    action: 'refs',
                    owner: null,
                    repository: null,
                    issue: '1100',
                    raw: ', #1100',
                    prefix: '#',
                },
            ]);
        });
    });
});
