// gitex-flow/build/test/GitFlowTester.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestBranch = exports.GitFlowTester = void 0;
const os_1 = require("os");
const chai_1 = require("chai");
const TestGitRepository_1 = require("./TestGitRepository");
const path_1 = require("path");
const fs_extra_1 = require("fs-extra");
/**
 * Tester for some standard git flow tests.
 */
class GitFlowTester {
    /**
     * Initializes a new instance of this class.
     *
     * @param gitFlow - Git flow implementation to be tested.
     * @param testRepoPath - Path the test repo is saved to.
     */
    constructor(gitFlow, testRepoPath) {
        this.gitFlow = gitFlow;
        this.repo = new TestGitRepository_1.TestGitRepository({
            projectPath: testRepoPath,
        });
    }
    /**
     * Init the git flow tester.
     */
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.dispose();
            yield GitFlowTester.ensureGitFlowRepo(this.gitFlow, this.repo);
        });
    }
    /**
     * Disposes the git flow tester.
     */
    dispose() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.repo.remove();
        });
    }
    /**
     * Clears the cache.
     */
    static clearCache() {
        return __awaiter(this, void 0, void 0, function* () {
            yield fs_extra_1.emptyDir(GitFlowTester.CachedGitFlowRepo);
            yield fs_extra_1.rmdir(GitFlowTester.CachedGitFlowRepo);
        });
    }
    /**
     * Asserts the version.
     */
    assertVersion() {
        return __awaiter(this, void 0, void 0, function* () {
            const version = yield this.gitFlow.version();
            chai_1.assert.match(version, /^[0-9]+.[0-9]+.[0-9]+.*/);
        });
    }
    /**
     * Asserts the initialization of the test git flow.
     *
     * @param config - The git flow configuration.
     * @param force - Flag to force the initialization.
     */
    assertInit(config, force) {
        var _a, _b, _c, _d, _e, _f, _g;
        return __awaiter(this, void 0, void 0, function* () {
            yield this.repo.ensure();
            yield this.gitFlow.init(config, force);
            const activeConfig = yield this.gitFlow.config.get();
            chai_1.assert.equal(activeConfig.masterBranch, (_a = config === null || config === void 0 ? void 0 : config.masterBranch) !== null && _a !== void 0 ? _a : 'master');
            chai_1.assert.equal(activeConfig.developBranch, (_b = config === null || config === void 0 ? void 0 : config.developBranch) !== null && _b !== void 0 ? _b : 'develop');
            chai_1.assert.equal(activeConfig.featureBranchPrefix, (_c = config === null || config === void 0 ? void 0 : config.featureBranchPrefix) !== null && _c !== void 0 ? _c : 'feature/');
            chai_1.assert.equal(activeConfig.bugfixBranchPrefix, (_d = config === null || config === void 0 ? void 0 : config.bugfixBranchPrefix) !== null && _d !== void 0 ? _d : 'bugfix/');
            chai_1.assert.equal(activeConfig.releaseBranchPrefix, (_e = config === null || config === void 0 ? void 0 : config.releaseBranchPrefix) !== null && _e !== void 0 ? _e : 'release/');
            chai_1.assert.equal(activeConfig.hotfixBranchPrefix, (_f = config === null || config === void 0 ? void 0 : config.hotfixBranchPrefix) !== null && _f !== void 0 ? _f : 'hotfix/');
            chai_1.assert.equal(activeConfig.supportBranchPrefix, (_g = config === null || config === void 0 ? void 0 : config.supportBranchPrefix) !== null && _g !== void 0 ? _g : 'support/');
            chai_1.assert.equal(activeConfig.versionTagPrefix, config === null || config === void 0 ? void 0 : config.versionTagPrefix);
        });
    }
    /**
     * Checks out the development branch.
     */
    checkoutDevelopBranch() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.checkoutBranch('develop');
        });
    }
    /**
     * Checks out a branch.
     *
     * @param branchName - The branch name to be checked out.
     */
    checkoutBranch(branchName) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.repo.checkout(branchName);
        });
    }
    /**
     * Select the given git flow branch.
     *
     * @param type - The git flow branch type to be selected.
     *
     * @returns The selected git flow branch.
     */
    selectBranch(type) {
        switch (type) {
            case 'bugfix':
                return new TestBranch(this.gitFlow.bugfix, this.repo);
            case 'feature':
                return new TestBranch(this.gitFlow.feature, this.repo);
            case 'hotfix':
                return new TestBranch(this.gitFlow.hotfix, this.repo);
            case 'release':
                return new TestBranch(this.gitFlow.release, this.repo);
            case 'support':
                return new TestBranch(this.gitFlow.support, this.repo);
            default:
                throw new Error(`Type "${type}" is not supported.`);
        }
    }
    static ensureGitFlowRepo(gitFlow, repo) {
        return __awaiter(this, void 0, void 0, function* () {
            const cachePath = GitFlowTester.CachedGitFlowRepo;
            if (yield fs_extra_1.pathExists(cachePath)) {
                yield fs_extra_1.copy(cachePath, repo.getRepoPath());
            }
            else {
                yield repo.commitTestFile('package.json', 'chore(project): Added package.json');
                yield gitFlow.init();
                yield fs_extra_1.ensureDir(cachePath);
                yield fs_extra_1.copy(repo.getRepoPath(), cachePath);
            }
        });
    }
}
exports.GitFlowTester = GitFlowTester;
GitFlowTester.CachedGitFlowRepo = path_1.join(os_1.tmpdir(), 'GitFlowRepoTemplate');
/**
 * A GitFlowBranch with some extended functionality for easier unit tests.
 */
class TestBranch {
    /**
     * Initializes a new instance of this class.
     *
     * @param branch - Branch to be wrapped.
     * @param repo - The test repo.
     */
    constructor(branch, repo) {
        this.branch = branch;
        this.repo = repo;
        this.type = branch.type;
        this.defaultBase = branch.defaultBase;
    }
    /**
     * Gets the test git flow branch configuration.
     *
     * @returns The git flow branch configuration.
     */
    getConfig() {
        return this.branch.getConfig();
    }
    /**
     * Retrieves all test  branches from the current type.
     *
     * @returns All open branches from the current type.
     */
    list() {
        return this.branch.list();
    }
    /**
     * Starts the current test git flow branch.
     *
     * @param name - The name of the git flow branch.
     * @param base - The base of the git flow branch.
     *
     * @returns The name with its prefix.
     */
    start(name, base) {
        return __awaiter(this, void 0, void 0, function* () {
            this.branchName = yield this.branch.start(name, base);
            return this.branchName;
        });
    }
    /**
     * Finishes the current test git flow branch.
     *
     * @param name - The name of the git flow branch.
     * @param msg - The message to finish the test git flow branch.
     */
    finish(name, msg) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.branch.finish(name, msg);
        });
    }
    /**
     * Generates the test git flow branch name.
     *
     * @param name - The name of the test git flow branch.
     *
     * @returns The test git flow branch name.
     */
    generateBranchName(name) {
        return __awaiter(this, void 0, void 0, function* () {
            return name;
        });
    }
    /**
     * Commits a given file to the test repository.
     *
     * @param fileName - The file name to be commited.
     * @param msg - The commit message.
     */
    commit(fileName, msg) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.branchName) {
                throw new Error('Branch was not started.');
            }
            yield this.repo.checkout(this.branchName);
            yield this.repo.commitTestFile(fileName, msg);
        });
    }
}
exports.TestBranch = TestBranch;
