// gitex-flow/build/test/KeepAChangelogWriter.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const KeepAChangelogWriter_1 = require("../src/changelog/KeepAChangelogWriter");
const Utils_1 = require("../src/tools/Utils");
const testTmpPath = path_1.resolve(path_1.join('.', 'test_tmp'));
const changelogPath = path_1.join(testTmpPath, 'CHANGELOG.md');
describe('Test Keep-A-Changelog writer implementation', function () {
    this.beforeAll(() => __awaiter(this, void 0, void 0, function* () {
        yield fs_extra_1.emptyDir(testTmpPath);
    }));
    this.afterAll(() => __awaiter(this, void 0, void 0, function* () {
        yield fs_extra_1.remove(testTmpPath);
    }));
    it('should create, parse and append changelog', function () {
        return __awaiter(this, void 0, void 0, function* () {
            const logs = yield Utils_1.Utils.parseConventionalCommits([
                `feat(test): Added feature test.txt

      closes #78`,
                `fix(test): Fixed errors in test.txt

      closes #80`,
                `fix(service): Removed support of unencrypted HTTP protocol

      This unencrypted protocol has led to several vulnerabilities in the framework.
      
      SECURITY: Only encrypted protocols are now allowed
      BREAKING CHANGE: Removed HTTP endpoint in web service.
      
      closes #941, refs #1094, #1100`,
                `Merge tag '1.4.4' into develop

      1.4.4 1.4.4`,
            ]);
            const writer = new KeepAChangelogWriter_1.KeepAChangelogWriter({
                basePath: testTmpPath,
            });
            yield writer.write({
                version: '1.0.0',
                title: 'Test project',
                repoUrl: 'https://github.com/gitex-flow/gitex-flow-node',
                date: '2021-03-16',
            }, [logs[0]]);
            chai_1.assert.equal(yield fs_extra_1.readFile(changelogPath, 'utf8'), yield fs_extra_1.readFile(path_1.resolve(path_1.join(__dirname, 'files', 'KaC_1_0_0.md')), 'utf8'));
            yield writer.write({
                version: '1.0.1',
                title: 'Test project',
                repoUrl: 'https://github.com/gitex-flow/gitex-flow-node',
                date: '2021-03-16',
            }, [logs[1]]);
            chai_1.assert.equal(yield fs_extra_1.readFile(changelogPath, 'utf8'), yield fs_extra_1.readFile(path_1.resolve(path_1.join(__dirname, 'files', 'KaC_1_1_0.md')), 'utf8'));
            yield writer.write({
                version: '2.0.0',
                title: 'Test project',
                repoUrl: 'https://github.com/gitex-flow/gitex-flow-node',
                date: '2021-03-16',
            }, [logs[2]]);
            chai_1.assert.equal(yield fs_extra_1.readFile(changelogPath, 'utf8'), yield fs_extra_1.readFile(path_1.resolve(path_1.join(__dirname, 'files', 'KaC_2_0_0.md')), 'utf8'));
            yield writer.write({
                version: '2.0.1',
                title: 'Test project',
                repoUrl: 'https://github.com/gitex-flow/gitex-flow-node',
                date: '2021-03-16',
            }, [logs[3]]);
            chai_1.assert.equal(yield fs_extra_1.readFile(changelogPath, 'utf8'), yield fs_extra_1.readFile(path_1.resolve(path_1.join(__dirname, 'files', 'KaC_2_0_1.md')), 'utf8'));
        });
    });
});
