// gitex-flow/build/src/gflow/branches/GFlowReleaseBranch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GFlowReleaseBranch = void 0;
const GitFlowNodeProject_1 = require("../../tools/GitFlowNodeProject");
const GFlowBranch_1 = require("./GFlowBranch");
const Utils_1 = require("../../tools/Utils");
/**
 * This class extending a release branch with some helpful functionality.
 */
class GFlowReleaseBranch extends GFlowBranch_1.GFlowBranch {
    /**
     * Initializes a new instance of this class.
     *
     * @param gitFlowBranch - Git flow branch to be wrapped.
     * @param options - Git flow node project options.
     */
    constructor(gitFlowBranch, options) {
        super(gitFlowBranch, options);
    }
    /**
     * Creates and starts a new release branch.
     *
     * @param name - Name of the branch to be started.
     * @param base - Base of the branch should be started from.
     *
     * @returns The name of the release branch.
     */
    start(name, base) {
        const _super = Object.create(null, {
            stashChanges: { get: () => super.stashChanges },
            start: { get: () => super.start }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const version = yield this.generateBranchName(name);
            if (!version) {
                throw new Error('Failed to calculate the version from the current repository.');
            }
            const project = new GitFlowNodeProject_1.GitFlowNodeProject(this.projectConfig);
            const stashed = yield _super.stashChanges.call(this, project);
            const branchName = yield _super.start.call(this, version, base);
            yield project.writeVersion(version);
            const changelogConfig = Utils_1.Utils.deriveChangelogConfig(this.projectConfig);
            yield project.updateChangelog(changelogConfig, version);
            yield project.commitChanges();
            if (stashed) {
                yield this.popStashedChanges(project);
            }
            return branchName;
        });
    }
    /**
     * Merges and finishes the branch of the branch type '[[type]]'.
     *
     * @param name - Name of the branch to be finished.
     * @param msg - Message to be set for finishing the branch.
     */
    finish(name, msg) {
        const _super = Object.create(null, {
            finish: { get: () => super.finish }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const project = new GitFlowNodeProject_1.GitFlowNodeProject(this.projectConfig);
            const version = yield this.getVersion(project, name);
            msg = msg !== null && msg !== void 0 ? msg : version;
            yield _super.finish.call(this, version, msg);
        });
    }
    getVersion(project, name) {
        return __awaiter(this, void 0, void 0, function* () {
            let version = name;
            if (!version) {
                const releaseBranches = yield this.list();
                if (releaseBranches.length == 0) {
                    version = yield project.getVersion();
                }
                // There is only one release branch
                version = releaseBranches[0];
            }
            return version;
        });
    }
}
exports.GFlowReleaseBranch = GFlowReleaseBranch;
