// gitex-flow/build/src/tools/GitFlowNodeProject.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitFlowNodeProject = void 0;
const path_1 = require("path");
const GitRepository_1 = require("../git/GitRepository");
const fs_extra_1 = require("fs-extra");
const Utils_1 = require("./Utils");
const write_json_file_1 = __importDefault(require("write-json-file"));
const log4js_1 = require("log4js");
const ChangelogWriter_1 = require("../changelog/ChangelogWriter");
const ChangelogWriterFactory_1 = require("../changelog/ChangelogWriterFactory");
/**
 * Representing an API for handling git flow SemVer.
 */
class GitFlowNodeProject {
    /**
     * Initializes a new instance of this class.
     *
     * @param options - Options of the git flow node project instance.
     */
    constructor(options) {
        var _a, _b, _c;
        this.logger = log4js_1.getLogger('GitFlowNodeProject');
        if (!options)
            options = { projectPath: process.cwd() };
        options.projectPath = (_a = options.projectPath) !== null && _a !== void 0 ? _a : process.cwd();
        this.options = options;
        this.options.versionFile = (_b = this.options.versionFile) !== null && _b !== void 0 ? _b : GitFlowNodeProject.DefaultVersionFile;
        this.options.bumpVersionFiles = (_c = options.bumpVersionFiles) !== null && _c !== void 0 ? _c : GitFlowNodeProject.DefaultBumpVersionFiles;
        this.options.changelog = Utils_1.Utils.deriveChangelogConfig(this.options);
        this.gitRepository = new GitRepository_1.GitRepository(options);
    }
    /**
     * Checks out the given branch of the project.
     *
     * @param branchName - Name of the branch to be checked out.
     */
    checkoutBranch(branchName) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.gitRepository.ensureNoUnCommitedChanges();
            yield this.gitRepository.checkout(branchName);
        });
    }
    /**
     * Stashes the uncommited changes from the current branch.
     *
     * @returns `true` if stash was successful, otherwise `false`.
     */
    stash() {
        return __awaiter(this, void 0, void 0, function* () {
            const status = yield this.gitRepository.status();
            let stashMessage = undefined;
            if (!status.isClean()) {
                stashMessage = yield this.gitRepository.stash();
            }
            return stashMessage !== undefined;
        });
    }
    /**
     * Pops the latest stash.
     *
     * @returns Promise on popping the latest stash.
     */
    popLatestStash() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.gitRepository.popLatestStash();
        });
    }
    /**
     * Gets the current branch.
     *
     * @returns The current branch checked out.
     */
    getCurrentBranch() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const status = yield this.gitRepository.status();
            return (_a = status.current) !== null && _a !== void 0 ? _a : '';
        });
    }
    /**
     *  Writes the version and commits the changes in the git repository.
     *
     * @param version - Version to commit.
     */
    writeVersion(version) {
        return __awaiter(this, void 0, void 0, function* () {
            const versionFiles = this.options.bumpVersionFiles;
            for (const versionFile of versionFiles) {
                yield this.writeVersionToFile(versionFile, version);
            }
            this.logger.info(`Updated versions in ${versionFiles.join(' and ')} to ${version}`);
        });
    }
    /**
     * Updates the changelog with the changes since the last release.
     *
     * @param changelogConfig - The changelog configuration.
     * @param version - Version the changelog is created for.
     * @param name - Name of the release.
     */
    updateChangelog(changelogConfig, version, name) {
        return __awaiter(this, void 0, void 0, function* () {
            const changelogWriter = ChangelogWriterFactory_1.ChangelogWriterFactory.create(changelogConfig);
            if (changelogWriter) {
                const logs = yield this.gitRepository.getLogsSinceLastRelease();
                version = version !== null && version !== void 0 ? version : (yield this.getVersion());
                const context = yield this.getContext(version, name);
                yield changelogWriter.write(context, logs);
            }
        });
    }
    /**
     * Commits the changes of the git repository.
     *
     * @param commitVersionFiles - Indicates if the defined version files should be committed if they exists.
     * @param commitChangelog - Indicates if the changelog should be committed.
     *
     * @returns The hash of the commit.
     */
    commitChanges(commitVersionFiles = true, commitChangelog = true) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const updateDescs = [];
            const files = [];
            if (commitVersionFiles) {
                updateDescs.push('version');
                const versionFiles = this.options.bumpVersionFiles;
                for (const versionFile of versionFiles) {
                    const versionFilePath = path_1.join(this.options.projectPath, versionFile);
                    if (yield fs_extra_1.pathExists(versionFilePath)) {
                        files.push(versionFile);
                    }
                }
            }
            if (commitChangelog) {
                updateDescs.push('changelog');
                const changelog = (_a = this.options.changelog) === null || _a === void 0 ? void 0 : _a.changelogFileName;
                const changelogPath = path_1.join(this.options.projectPath, changelog);
                if (yield fs_extra_1.pathExists(changelogPath)) {
                    files.push(changelog);
                }
                const latestChangelogName = ChangelogWriter_1.ChangelogWriter.getLatestChangelogName(changelog);
                const latestChangelogPath = path_1.join(this.options.projectPath, latestChangelogName);
                if (yield fs_extra_1.pathExists(latestChangelogPath)) {
                    files.push(latestChangelogPath);
                }
            }
            const commitMsg = `chore(release): Updated ${updateDescs.join(' and ')}`;
            return this.gitRepository.commit(files, commitMsg);
        });
    }
    /**
     * Gets the current version from the package.json.
     *
     * @returns The version of the project.
     */
    getVersion() {
        return __awaiter(this, void 0, void 0, function* () {
            const versionFile = this.options.versionFile;
            const packageJson = yield fs_extra_1.readJson(path_1.join(this.options.projectPath, versionFile));
            return packageJson.version;
        });
    }
    writeVersionToFile(fileName, version) {
        return __awaiter(this, void 0, void 0, function* () {
            const filePath = path_1.join(this.options.projectPath, fileName);
            if (yield fs_extra_1.pathExists(filePath)) {
                const packageJson = yield fs_extra_1.readJson(filePath);
                packageJson.version = version;
                yield write_json_file_1.default(filePath, packageJson, { detectIndent: true });
            }
        });
    }
    getContext(version, name) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const versionFile = this.options.versionFile;
            const packageJson = yield fs_extra_1.readJson(path_1.join(this.options.projectPath, versionFile));
            const repoUrl = ((_a = packageJson === null || packageJson === void 0 ? void 0 : packageJson.repository) === null || _a === void 0 ? void 0 : _a.url) ? new URL(packageJson.repository.url) : undefined;
            const host = repoUrl ? `https://${repoUrl.host}` : undefined;
            let path = repoUrl === null || repoUrl === void 0 ? void 0 : repoUrl.pathname;
            if (path === null || path === void 0 ? void 0 : path.endsWith('.git')) {
                path = path.substring(0, path.length - 4);
            }
            const url = repoUrl ? `https://${repoUrl.host}${path}` : undefined;
            return {
                version: version,
                title: name,
                host: host,
                repoUrl: url,
                commit: 'commits',
                issue: 'issues',
                date: Utils_1.Utils.getCurrDate(),
            };
        });
    }
}
exports.GitFlowNodeProject = GitFlowNodeProject;
GitFlowNodeProject.DefaultVersionFile = 'package.json';
GitFlowNodeProject.DefaultBumpVersionFiles = [GitFlowNodeProject.DefaultVersionFile, 'package-lock.json'];
