// gitex-flow/build/src/tools/Utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const log4js_1 = require("log4js");
const stream_1 = require("stream");
const ChangelogType_1 = require("../changelog/ChangelogType");
const changelog_1 = require("../changelog");
const conventional_commits_parser_1 = __importDefault(require("conventional-commits-parser"));
/**
 * Provides some utility functions.
 */
class Utils {
    /**
     * Gets the current date formatted as yyyy-mm-dd.
     *
     * @returns date in fomat yyyy-mm-dd.
     */
    static getCurrDate() {
        const today = new Date();
        const yyyy = today.getFullYear();
        let dd = '' + today.getDate();
        let mm = '' + (today.getMonth() + 1);
        if (+dd < 10) {
            dd = '0' + dd;
        }
        if (+mm < 10) {
            mm = '0' + mm;
        }
        return `${yyyy}-${mm}-${dd}`;
    }
    /**
     * Executes a command and suppresses errors if they are thrown.
     *
     * @param command - Command to be executed.
     */
    static exec(command) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield command();
            }
            catch (error) {
                const logger = log4js_1.getLogger('gitex-flow');
                logger.error(error);
            }
        });
    }
    /**
     * Prints the config to the console.
     *
     * @param gitFlow - The git flow instance the config should be printed.
     */
    static printConfig(gitFlow) {
        return __awaiter(this, void 0, void 0, function* () {
            const branches = yield gitFlow.config.get();
            console.info(branches);
        });
    }
    /**
     * Prints the branches to the console.
     *
     * @param gitFlowBranch - The branch type to be printed.
     */
    static printBranches(gitFlowBranch) {
        return __awaiter(this, void 0, void 0, function* () {
            const branches = yield gitFlowBranch.list();
            if (branches.length === 0) {
                console.error(`There are no active ${gitFlowBranch.type} branches.`);
            }
            else {
                console.info(`Active ${gitFlowBranch.type} branches:`);
                for (const branch of branches) {
                    console.info(` - ${branch}`);
                }
            }
        });
    }
    /**
     * Pipes a readable stream asynchrounously to a writable stream with error handling.
     *
     * @param src - The readable source stream.
     * @param dest - The writable destination stream.
     * @param destroy - Specifies if the streams should be destroyed on finish.
     *
     * @returns Promise on copying stream properly.
     */
    static pipe(src, dest, destroy = true) {
        return new Promise((resolve, reject) => {
            src
                .on('error', (err) => {
                err.message = `Error on reading source stream: ${err.message}`;
                reject(err);
            })
                .pipe(dest)
                .on('error', (err) => {
                err.message = `Error on writing destination stream: ${err.message}`;
                reject(err);
            })
                .on('finish', () => {
                if (destroy) {
                    src.destroy();
                    dest.destroy();
                }
                resolve();
            });
        });
    }
    /**
     * Derives the [[ChangelogConfig]] from a given [[projectConfig]].
     *
     * @param projectConfig - The project configuration.
     *
     * @returns The derived changelog config.
     */
    static deriveChangelogConfig(projectConfig) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const config = (_a = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.changelog) !== null && _a !== void 0 ? _a : {
            basePath: (_b = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.projectPath) !== null && _b !== void 0 ? _b : process.cwd(),
            type: ChangelogType_1.ChangelogType.ConventionalChangelog,
        };
        config.basePath = (_d = (_c = config.basePath) !== null && _c !== void 0 ? _c : projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.projectPath) !== null && _d !== void 0 ? _d : process.cwd();
        // Following lines are ensuring backward compatibility to avoid a breaking change for version 2.3.
        config.changelogFileName =
            (_f = (_e = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.changelogFileName) !== null && _e !== void 0 ? _e : config.changelogFileName) !== null && _f !== void 0 ? _f : changelog_1.ChangelogWriter.DefaultChangelogFile;
        config.storeLatestChangelog = (_g = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.storeLatestChangelog) !== null && _g !== void 0 ? _g : config.storeLatestChangelog;
        if (config.type == ChangelogType_1.ChangelogType.ConventionalChangelog) {
            const conf = config;
            conf.conventionalChangelogPresent =
                (_j = (_h = projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.conventionalChangelogPresent) !== null && _h !== void 0 ? _h : conf.conventionalChangelogPresent) !== null && _j !== void 0 ? _j : 'angular';
        }
        return config;
    }
    /**
     * Parses conventional commit messages to a [[GitLog]] array.
     *
     * @param commitMessages - The commit messages.
     * @param conventionalCommitConfig - The configuration of the conventional commit parser.
     * @returns The parsed conventional commit messages as an array of [[GitLogs]].
     */
    static parseConventionalCommits(commitMessages, conventionalCommitConfig) {
        return __awaiter(this, void 0, void 0, function* () {
            const gitLogs = [];
            const stream = stream_1.Readable.from(commitMessages);
            return new Promise((resolve, reject) => {
                stream
                    .pipe(Utils.parseConventionalCommitsViaPipe(conventionalCommitConfig))
                    .on('error', function (err) {
                    err.message = 'Error in conventional-commits-parser: ' + err.message;
                    reject(err);
                })
                    .on('data', (data) => {
                    const log = data;
                    gitLogs.push(log);
                })
                    .on('end', () => {
                    resolve(gitLogs);
                });
            });
        });
    }
    /**
     * Parses conventional commit messages via a stream.Transform pipe.
     *
     * @param conventionalCommitConfig - The configuration of the conventional commit parser.
     * @returns The parsed conventional commit messages as transformed stream.
     */
    static parseConventionalCommitsViaPipe(conventionalCommitConfig) {
        return conventional_commits_parser_1.default(conventionalCommitConfig !== null && conventionalCommitConfig !== void 0 ? conventionalCommitConfig : {
            referenceActions: [
                'close',
                'closes',
                'closed',
                'fix',
                'fixes',
                'fixed',
                'resolve',
                'resolves',
                'resolved',
                'refs',
                'references',
            ],
            noteKeywords: ['BREAKING CHANGE', 'SECURITY', 'REMOVED'],
        });
    }
}
exports.Utils = Utils;
