// gitex-flow/build/src/avh/branches/AvhGitFlowBranch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvhGitFlowBranch = void 0;
const log4js_1 = require("log4js");
const AvhBranchListParser_1 = require("../AvhBranchListParser");
const GitFlowBashExecuter_1 = require("../GitFlowBashExecuter");
/**
 * This class implements the basic functionality of a git flow branch.
 */
class AvhGitFlowBranch {
    /**
     * Initializes a new instance of this class.
     *
     * @param repoPath - The path to the git repository.
     */
    constructor(repoPath) {
        this.logger = log4js_1.getLogger('AvhGitFlowBranch');
        this.repositoryPath = repoPath;
    }
    /**
     * {@inheritdoc}
     *
     * @returns The list of the currently opened branch.
     */
    list() {
        return __awaiter(this, void 0, void 0, function* () {
            const output = yield GitFlowBashExecuter_1.GitFlowBashExecuter.execute({
                type: this.type,
                action: 'list',
                repositoryPath: this.repositoryPath,
            });
            return AvhBranchListParser_1.AvhBranchListParser.parse(output);
        });
    }
    /**
     * {@inheritdoc}
     *
     * @param name - Name of the branch to be started.
     * @param base - Base of the branch should be started from.
     * @returns The git reference of the create branch.
     */
    start(name, base) {
        return __awaiter(this, void 0, void 0, function* () {
            let branchName = yield this.getBranchNameFromConfig(name);
            const output = yield GitFlowBashExecuter_1.GitFlowBashExecuter.execute({
                type: this.type,
                action: 'start',
                name: name,
                repositoryPath: this.repositoryPath,
                args: base ? [base] : undefined,
            });
            const matches = output.match(/'([^']+)'/);
            if (matches && matches.groups) {
                if (branchName !== matches.groups[0]) {
                    this.logger.warn(`WARNING: The expected branch name "${branchName}" does not match the actual branch name '${matches.groups[0]}'`);
                }
                branchName = matches.groups[0];
            }
            return branchName;
        });
    }
    /**
     * {@inheritdoc}
     *
     * @param name - Name of the branch to be finished.
     * @param msg - Message to be set for finishing the branch.
     */
    finish(name, msg) {
        return __awaiter(this, void 0, void 0, function* () {
            let options = undefined;
            if (msg) {
                options = `-m ${msg}`;
            }
            const output = yield GitFlowBashExecuter_1.GitFlowBashExecuter.execute({
                type: this.type,
                action: 'finish',
                name: name,
                repositoryPath: this.repositoryPath,
                options: options,
            });
            const outputs = output.trim().split('\n');
            for (const out of outputs) {
                this.logger.info(out);
            }
        });
    }
    /**
     * {@inheritdoc}
     *
     * @param name - A custom name for the branch.
     *
     * @returns The generated branch name.
     */
    generateBranchName(name) {
        return __awaiter(this, void 0, void 0, function* () {
            return name;
        });
    }
    getBranchNameFromConfig(name) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const config = yield this.getConfig();
            let prefix = (_a = config.prefix) !== null && _a !== void 0 ? _a : this.type;
            if (prefix.endsWith('/')) {
                prefix = prefix.slice(0, -1);
            }
            return `${prefix}/${name}`;
        });
    }
}
exports.AvhGitFlowBranch = AvhGitFlowBranch;
