// gitex-flow/build/src/avh/branches/FeatureGitFlowBranch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureGitFlowBranch = void 0;
const AvhGitFlowBranch_1 = require("./AvhGitFlowBranch");
/**
 * This class wraps the feature branch of the AVH implementation.
 */
class FeatureGitFlowBranch extends AvhGitFlowBranch_1.AvhGitFlowBranch {
    /**
     * Initializes a new instance of this class.
     *
     * @param repoPath - The path to the git repository.
     * @param configProvider - Git flow config provider.
     */
    constructor(repoPath, configProvider) {
        super(repoPath);
        /**
         * {@inheritdoc}
         */
        this.type = 'feature';
        /**
         * {@inheritdoc}
         */
        this.defaultBase = 'develop';
        this.configProvider = configProvider;
    }
    /**
     * {@inheritdoc}
     *
     * @returns The configuration of the feature git flow branch.
     */
    getConfig() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const config = yield ((_a = this.configProvider) === null || _a === void 0 ? void 0 : _a.get());
            return {
                prefix: config === null || config === void 0 ? void 0 : config.featureBranchPrefix,
            };
        });
    }
}
exports.FeatureGitFlowBranch = FeatureGitFlowBranch;
