// gitex-flow/build/src/git/GitRepository.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitRepository = void 0;
const simple_git_1 = __importDefault(require("simple-git"));
const path_1 = require("path");
const fs_extra_1 = require("fs-extra");
const tools_1 = require("../tools");
const semver_1 = require("semver");
/**
 * A simple API with basic functionality of a git repository.
 */
class GitRepository {
    /**
     * Initializes a new instance of this class.
     *
     * @param config - The project configuration.
     */
    constructor(config) {
        this.config = config;
    }
    /**
     * Gets the folder path of the git repository.
     *
     * @returns The path to the git repository.
     */
    getRepoPath() {
        var _a, _b;
        return (_b = (_a = this.config) === null || _a === void 0 ? void 0 : _a.projectPath) !== null && _b !== void 0 ? _b : process.cwd();
    }
    /**
     * Gets the path of the git repository.
     */
    remove() {
        return __awaiter(this, void 0, void 0, function* () {
            yield fs_extra_1.emptyDir(this.getRepoPath());
            yield fs_extra_1.rmdir(this.getRepoPath());
        });
    }
    /**
     * Ensures the repository exists.
     * If it doesn't exist it will be created.
     */
    ensure() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.createOrOpenRepo();
        });
    }
    /**
     * Checks out a given branch.
     *
     * @param branchName - Name of the branch to be checked out.
     */
    checkout(branchName) {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            yield repo.checkout(branchName);
        });
    }
    /**
     * Retrieves the current status of the git repository.
     *
     * @returns The status of the git repository.
     */
    status() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            const status = yield repo.status();
            return status;
        });
    }
    /**
     * Adds and commits the given file names to the current branch.
     *
     * @param fileNames - Relative file paths to be added before commit.
     * @param message - Commit message.
     * @param authorName - The name of the author.
     * @param authorMail - Mail address of the author.
     *
     * @returns The hash of the commit.
     */
    commit(fileNames, message, authorName, authorMail) {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            for (const fileName of fileNames) {
                yield repo.add(fileName);
            }
            let options = undefined;
            if (authorName && authorMail) {
                options = { '--author': `"${authorName} <${authorMail}>"` };
            }
            const hash = yield repo.commit(message, fileNames, options);
            return hash.commit;
        });
    }
    /**
     * Stashes the uncommited changes from the current branch.
     *
     * @returns The message of the stashing.
     */
    stash() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            const message = yield repo.stash();
            return message.trim();
        });
    }
    /**
     * Pops stash with a given name.
     */
    popLatestStash() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            yield repo.stash(['pop', '-q']);
        });
    }
    /**
     * Ensures there are no uncommited changes (staged and unstaged) in the local workspace.
     */
    ensureNoUnCommitedChanges() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            const diff = yield repo.diff(['HEAD']);
            if (diff) {
                throw new Error('There are some uncommited changes.');
            }
        });
    }
    /**
     * Returns the most recent released version tag (semantic version).
     *
     * @returns The version of the latest release.
     */
    getLatestReleasedVersion() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            const gitTags = yield repo.tags();
            // Filter all prerelease tags
            const tags = gitTags.all.filter((tag) => {
                const pre = semver_1.prerelease(tag);
                return !pre;
            });
            return tags.pop();
        });
    }
    /**
     * Collects all commit messages since the last release.
     *
     * @returns The logs since the last release.
     */
    getLogsSinceLastRelease() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const logs = yield this.getDiffLogs();
            const logMessages = logs.map((log) => `${log.message}\n\n${log.body}\n\n${log.refs}`);
            const gitLogs = yield tools_1.Utils.parseConventionalCommits(logMessages, (_a = this.config) === null || _a === void 0 ? void 0 : _a.conventionalCommit);
            gitLogs.forEach((v, i) => (v.hash = logs[i].hash));
            return gitLogs;
        });
    }
    getDiffLogs() {
        return __awaiter(this, void 0, void 0, function* () {
            const repo = yield this.createOrOpenRepo();
            const latestVersion = yield this.getLatestReleasedVersion();
            const logs = yield repo.log({
                from: 'HEAD',
                to: latestVersion,
                symmetric: true,
            });
            return logs.all;
        });
    }
    createOrOpenRepo() {
        return __awaiter(this, void 0, void 0, function* () {
            let repo;
            const path = this.getRepoPath();
            const gitFolder = path_1.join(path, '.git');
            yield fs_extra_1.ensureDir(path);
            if (yield fs_extra_1.pathExists(gitFolder)) {
                repo = simple_git_1.default(path);
            }
            else {
                repo = simple_git_1.default(path);
                yield repo.init();
            }
            return repo;
        });
    }
}
exports.GitRepository = GitRepository;
