// gitex-flow/build/src/gflow/GFlow.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GFlow = void 0;
const GFlowReleaseBranch_1 = require("./branches/GFlowReleaseBranch");
const GFlowHotFixBranch_1 = require("./branches/GFlowHotFixBranch");
const log4js_1 = require("log4js");
const GFlowBranch_1 = require("./branches/GFlowBranch");
/**
 * GitFlow wrapper extending functionality to a common git flow implementation.
 */
class GFlow {
    /**
     * Initializes a new instance of this class.
     *
     * @param gitFlow - GitFlow implementation.
     * @param options - Options for configuring the GFlow.
     */
    constructor(gitFlow, options) {
        options = this.ensureDefaults(options);
        if (options.log4jsConfig) {
            log4js_1.configure(options.log4jsConfig);
        }
        this.gitFlow = gitFlow;
        this.options = options;
        this.feature = new GFlowBranch_1.GFlowBranch(this.gitFlow.feature, options.projectConfig);
        this.bugfix = new GFlowBranch_1.GFlowBranch(this.gitFlow.bugfix, options.projectConfig);
        this.release = new GFlowReleaseBranch_1.GFlowReleaseBranch(this.gitFlow.release, options.projectConfig);
        this.hotfix = new GFlowHotFixBranch_1.GFlowHotFixBranch(this.gitFlow.hotfix, options.projectConfig);
        this.support = new GFlowBranch_1.GFlowBranch(this.gitFlow.support, options.projectConfig);
        this.config = this.gitFlow.config;
    }
    /**
     * Setup a git repository for git flow usage.
     *
     * @param config - The git flow configuration.
     * @param force - Force reinitialisation if git flow already initialized.
     */
    init(config, force) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.gitFlow.init(config !== null && config !== void 0 ? config : this.options.gitFlowConfig, force);
        });
    }
    /**
     * Provides the version of the git flow implementation.
     *
     * @returns The version of git flow.
     */
    version() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.gitFlow.version();
        });
    }
    ensureDefaults(options) {
        options = options !== null && options !== void 0 ? options : {};
        if (!options.projectConfig) {
            options.projectConfig = {
                projectPath: process.cwd(),
            };
        }
        if (!options.log4jsConfig) {
            options.log4jsConfig = {
                appenders: { console: { type: 'console' } },
                categories: { default: { appenders: ['console'], level: 'info' } },
            };
        }
        return options;
    }
}
exports.GFlow = GFlow;
